File: sol/solution/0001-0100/0033. Search in Rotated Sorted Array/Solution.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search(nums, target)
  l, r = 0, nums.length - 1
  while l <= r
    mid = (l + r) / 2
    return mid if nums[mid] == target
    if nums[l] <= nums[mid]
      if nums[l] <= target && target < nums[mid]
        r = mid - 1
      else
        l = mid + 1
      end
    else
      if nums[mid] < target && target <= nums[r]
        l = mid + 1
      else
        r = mid - 1
      end
    end
  end
  -1
end

File: sol/solution/0001-0100/0033. Search in Rotated Sorted Array/Solution.erl
-spec search([integer()], integer()) -> integer().
search(Nums, Target) ->
    search(Nums, Target, 0, length(Nums) - 1).
search(Nums, Target, L, R) when L =< R ->
    Mid = (L + R) div 2,
    Val = lists:nth(Mid+1, Nums),
    LVal = lists:nth(L+1, Nums),
    RVal = lists:nth(R+1, Nums),
    if Val =:= Target -> Mid;
       LVal =< Val ->
           if LVal =< Target, Target < Val ->
                  search(Nums, Target, L, Mid - 1);
              true -> search(Nums, Target, Mid + 1, R)
           end;
       true ->
           if Val < Target, Target =< RVal ->
                  search(Nums, Target, Mid + 1, R);
              true -> search(Nums, Target, L, Mid - 1)
           end
    end;
search(_, _, _, _) -> -1.